sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -180,6 +180,70 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
   return httpRequest;
 }
+
+static void CleanupThread()
+{
+  using namespace Azure::Core::Http::_detail;
+  for (;;)
+  {
+    {
+      // Won't continue until the ConnectionPoolMutex is released from MoveConnectionBackToPool
+      std::unique_lock<std::mutex> lockForPoolCleaning(
+          CurlConnectionPool::g_curlConnectionPool.ConnectionPoolMutex);
+      // Wait the defined default time OR to the signal from the conditional variable.
+      // wait_until releases the mutex lock until it wakes up again or it's cancelled.
+      if (CurlConnectionPool::g_curlConnectionPool.ConditionalVariableForCleanThread.wait_for(
+              lockForPoolCleaning,
+              std::chrono::milliseconds(DefaultCleanerIntervalMilliseconds),
+              []() { return CurlConnectionPool::g_curlConnectionPool.ConnectionCounter == 0; }))
+      {
+        // Cancelled by another thead or no connections on wakeup
+        return;
+      }
+
+      // loop the connection pool index - Note: lock is re-taken for the mutex
+      for (auto index = CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.begin();
+           index != CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.end();
+           index++)
+      {
+        if (index->second.size() == 0)
+        {
+          // Move the next pool index
+          continue;
+        }
+
+        // Pool index with waiting connections. Loop the connection pool backwards until
+        // a connection that is not expired is found or until all connections are removed.
+        for (auto connection = index->second.end();;)
+        {
+          // loop starts at end(), go back to previous possition. We know the list is
+          // size() > 0 so we are safe to go end() - 1 and find the last element in the
+          // list
+          connection--;
+          if (connection->get()->IsExpired())
```

Comment:
By doing this decrement here, aren't we always skipping the last connection in the pool list? Shouldn't the decrement happen in the for-loop signature, i.e. at the end of the loop?

Reply:
remember that calling list.end() gives a sentinel for the list EOF which is probably pointing to NULL

Reply:
Got it, thanks.

Reply:
Correct me if I am mistaken, but couldn't this code be written more simply as follows (I omitted the comments to show the logic simply, but please add them back):
```C++
for (auto connection = index->second.end() - 1; connection->get()->IsExpired(); connection--)
{
  connection = index->second.erase(connection);
  CurlConnectionPool::g_curlConnectionPool.ConnectionCounter -= 1;

  if (index->second.size() == 0)
  {
    break;
  }
}
```

Is there anything different in the logic between the above and what you have?

Maybe even merge the loop-terminating condition even further, like this:
```C++
for (auto connection = index->second.end() - 1; connection->get()->IsExpired() && index->second.size() != 0; connection--)
{
  connection = index->second.erase(connection);
  CurlConnectionPool::g_curlConnectionPool.ConnectionCounter -= 1;

}
```

Reply:
I think that can work, yes.  That's a good recommendation.